src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+                initialListOfItemsToBeIncluded,
+                highlightedList,
+                hasSuggestedItemOptions);
+        }
+
+        private static bool IsAfterDot(ITextSnapshot snapshot, ITrackingSpan applicableToSpan)
+        {
+            var position = applicableToSpan.GetStartPoint(snapshot).Position;
+            return position > 0 && snapshot[position - 1] == '.';
+        }
+
+        private AsyncCompletionData.FilteredCompletionModel HandleNormalFiltering(
+            Func<ImmutableArray<RoslynCompletionItem>, string, ImmutableArray<RoslynCompletionItem>> filterMethod,
+            string filterText,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters,
+            CompletionTriggerKind initialRoslynTriggerKind,
+            CompletionFilterReason filterReason,
+            char typeChar,
+            List<ExtendedFilterResult> itemsInList,
+            ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight> highlightedList,
+            bool hasSuggestedItemOptions)
+        {
+            // Not deletion.  Defer to the language to decide which item it thinks best
+            // matches the text typed so far.
+
+            // Ask the language to determine which of the *matched* items it wants to select.
+            var matchingItems = itemsInList.Where(r => r.FilterResult.MatchedFilterText)
+                                           .Select(t => t.FilterResult.CompletionItem)
+                                           .AsImmutable();
+
+            var chosenItems = filterMethod(matchingItems, filterText);
+
+            var recentItems = _recentItems;
+
+            // Of the items the service returned, pick the one most recently committed
+            var bestItem = Session.GetBestCompletionItemBasedOnMRU(chosenItems, recentItems);
+            VSCompletionItem uniqueItem = null;
+            int selectedItemIndex = 0;
+
+            // Determine if we should consider this item 'unique' or not.  A unique item
+            // will be automatically committed if the user hits the 'invoke completion' 
+            // without bringing up the completion list.  An item is unique if it was the
+            // only item to match the text typed so far, and there was at least some text
+            // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+            // like "WriteLine" since no filter text has actually been provided.  HOwever,
+            // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+            if (bestItem != null)
+            {
+                selectedItemIndex = itemsInList.IndexOf(i => Equals(i.FilterResult.CompletionItem, bestItem));
+                if (selectedItemIndex > -1 && bestItem != null && matchingItems.Length == 1 && filterText.Length > 0)
+                {
+                    uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+                }
+            }
```

Comment:
this logic seems to be new.  where did it come from?

Reply:
Some code came from: [https://github.com/dotnet/roslyn/blob/ca0e683cb40bd126b40088f5b44ef68092797ab7/src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs#L275](https://github.com/dotnet/roslyn/blob/ca0e683cb40bd126b40088f5b44ef68092797ab7/src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs#L275)
Some part could be new.



---
In reply to: [232510290](https://github.com/dotnet/roslyn/pull/29016#discussion_r232510290) [](ancestors = 232510290)